cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/common/DataprocUtilsTest.java

```diff
@@ -30,58 +49,147 @@ public class DataprocUtilsTest {
   }
 
   @Test
   public void testParseLabelsIgnoresWhitespace() {
     Map<String, String> expected = new HashMap<>();
     expected.put("k1", "v1");
     expected.put("k2", "v2");
-    Assert.assertEquals(expected,
-        DataprocUtils.parseLabels(" k1  =\tv1  ,\nk2 = v2  ", ",", "="));
+    assertEquals(expected,
+                 DataprocUtils.parseLabels(" k1  =\tv1  ,\nk2 = v2  ", ",", "="));
   }
 
   @Test
   public void testParseLabelsWithoutVal() {
     Map<String, String> expected = new HashMap<>();
     expected.put("k1", "");
     expected.put("k2", "");
-    Assert.assertEquals(expected, DataprocUtils.parseLabels("k1,k2=", ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels("k1,k2=", ",", "="));
   }
 
   @Test
   public void testParseLabelsIgnoresInvalidKey() {
     Map<String, String> expected = new HashMap<>();
-    Assert.assertEquals(expected, DataprocUtils.parseLabels(("A"), ",", "="));
-    Assert.assertEquals(expected, DataprocUtils.parseLabels(("0"), ",", "="));
-    Assert.assertEquals(expected, DataprocUtils.parseLabels(("a.b"), ",", "="));
-    Assert.assertEquals(expected, DataprocUtils.parseLabels(("a^b"), ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels(("A"), ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels(("0"), ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels(("a.b"), ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels(("a^b"), ",", "="));
     StringBuilder longStr = new StringBuilder();
     for (int i = 0; i < 64; i++) {
       longStr.append('a');
     }
-    Assert.assertEquals(expected, DataprocUtils.parseLabels(longStr.toString(), ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels(longStr.toString(), ",", "="));
   }
 
   @Test
   public void testParseLabelsIgnoresInvalidVal() {
     Map<String, String> expected = new HashMap<>();
-    Assert.assertEquals(expected, DataprocUtils.parseLabels("a=A", ",", "="));
-    Assert.assertEquals(expected, DataprocUtils.parseLabels("a=ab.c", ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels("a=A", ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels("a=ab.c", ",", "="));
     StringBuilder longStr = new StringBuilder();
     for (int i = 0; i < 64; i++) {
       longStr.append('a');
     }
-    Assert.assertEquals(expected, DataprocUtils.parseLabels(String.format("a=%s", longStr.toString()), ",", "="));
+    assertEquals(expected, DataprocUtils.parseLabels(String.format("a=%s", longStr.toString()), ",", "="));
+  }
+
+  @Test
+  public void testGetMetadataSuccessOnFirstAttempt() throws Exception {
+    // Arrange
+    String expectedMetadata = "Sample metadata";
+    HttpURLConnection mockConnection = mock(HttpURLConnection.class);
+    when(mockConnection.getInputStream()).thenReturn(new ByteArrayInputStream(expectedMetadata.getBytes()));
+
+    URL mockUrl = PowerMockito.mock(URL.class);
```

Comment:
Use of `PowerMockito` should be avoided because the test now become tightly coupled to the implementation details of the class.

Reply:
In this scenario, I explored the options but there isn't any simple way of avoiding using PowerMockito because we need to mock URL.class and final classes can only be mocked by PowerMockito.

Reply:
We can introduce a `ConnectionProvider` which can be mocked for providing your mock connection in the test, for example: https://github.com/cdapio/cdap/blob/4a71b42a6f24cc21609613154555b1aac467f0e9/cdap-credential-ext-gcp-wi/src/test/java/io/cdap/cdap/security/spi/credential/GcpWorkloadIdentityCredentialProviderTest.java#L205-L218

Reply:
Addressed in the new PR: https://github.com/cdapio/cdap/pull/15550